pkg/controller/controller_test.go

```diff
@@ -0,0 +1,238 @@
+	host                     ControllerHost
+	pvSource                 *framework.FakeControllerSource
+	pvcSource                *framework.FakeControllerSource
+	expectedClaimAddedCount  int
+	expectedVolumeAddedCount int
+	actualClaimAddedCount    int
+	actualClaimDeletedCount  int
+	actualVolumeAddedCount   int
+	actualVolumeDeletedCount int
+	lock                     sync.Mutex
+	uids                     map[string]interface{}
+}
+
+func NewTestController(pvCount, pvcCount int) *testControllerPlugin {
+	return &testControllerPlugin{
+		expectedVolumeAddedCount: pvCount,
+		expectedClaimAddedCount:  pvCount,
+		actualClaimAddedCount:    0,
+		actualClaimDeletedCount:  0,
+		actualVolumeAddedCount:   0,
+		actualVolumeDeletedCount: 0,
+		uids: make(map[string]interface{}),
+	}
+}
+
+func (plugin *testControllerPlugin) Init(host ControllerHost) {
+	plugin.host = host
+}
+
+func (plugin *testControllerPlugin) Name() string {
+	return "test-controller-plugin"
+}
+
+func (plugin *testControllerPlugin) DefaultSyncPeriod() time.Duration {
+	return 1 * time.Second
+}
+
+func (plugin *testControllerPlugin) CreateController() Controller {
+	return plugin
+}
+
+func (plugin *testControllerPlugin) Configs() []*framework.Config {
+	// source simulates an apiserver object endpoint.
+	plugin.pvSource = framework.NewFakeControllerSource()
+	plugin.pvcSource = framework.NewFakeControllerSource()
+
+	// This will hold the downstream state, as we know it.
+	pvDownstream := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	pvcDownstream := cache.NewStore(cache.MetaNamespaceKeyFunc)
+
+	// This will hold incoming changes. Note how we pass downstream in as a
+	// KeyLister, that way resync operations will result in the correct set
+	// of update/delete deltas.
+	pvDeltaFifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, pvDownstream)
+	pvcDeltaFifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, pvcDownstream)
+
+	pvConfig := &framework.Config{
+		Queue:            pvDeltaFifo,
+		ListerWatcher:    plugin.pvSource,
+		ObjectType:       &api.PersistentVolume{},
+		FullResyncPeriod: 5 * time.Second,
+		RetryOnError:     true,
+		Process:          plugin.processPersistentVolume,
+	}
+
+	pvcConfig := &framework.Config{
+		Queue:            pvcDeltaFifo,
+		ListerWatcher:    plugin.pvcSource,
+		ObjectType:       &api.PersistentVolumeClaim{},
+		FullResyncPeriod: 5 * time.Second,
+		RetryOnError:     true,
+		Process:          plugin.processPersistentVolumeClaim,
+	}
+
+	return []*framework.Config{pvConfig, pvcConfig}
+}
+
+func (plugin *testControllerPlugin) processPersistentVolume(obj interface{}) error {
+	plugin.lock.Lock()
+	defer plugin.lock.Unlock()
+
+	delta := obj.(cache.Deltas).Oldest()
+	pv := delta.Object.(*api.PersistentVolume)
+	if (delta.Type == cache.Added || delta.Type == cache.Sync) && !plugin.exists(pv.Name) {
+		plugin.actualVolumeAddedCount++
+		fmt.Println(delta.Type, pv.Name, plugin.actualVolumeAddedCount)
```

Comment:
debug code, I assume?  I like to flag debug code with a trailing //FIXME so I know it's debug